### PR TITLE
[PF-127] Don't rollback creation flight if resource is READY

### PR DESCRIPTION
**Pull request by yonghaoy**

### CODE REVIEW

File: src/main/java/bio/terra/rbs/service/resource/FlightMapKeys.java [hunk 1]
```diff
@@ -7,5 +7,5 @@
   public static final String CLOUD_RESOURCE_UID = "cloudResourceUid";
   public static final String GOOGLE_PROJECT_ID = "googleProjectId";
   public static final String RESOURCE_CONFIG = "resourceConfig";
-  public static final String RESOURCE_ID = "resourceId";
+  public static final String RESOURCE_READY = "resourceReady";
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/service/resource/FlightMapKeys.java:10]: Worth a comment on what this signifies.

File: src/main/java/bio/terra/rbs/service/resource/flight/CreateProjectStep.java [hunk 2]
```diff
@@ -53,8 +54,14 @@ public StepResult doStep(FlightContext flightContext) throws RetryException {
 
   @Override
   public StepResult undoStep(FlightContext flightContext) {
+    FlightMap workingMap = flightContext.getWorkingMap();
+    // Don't do anything if resource is READY.
+    if (workingMap.get(RESOURCE_READY, Boolean.class) != null
+        && workingMap.get(RESOURCE_READY, Boolean.class)) {
+      return StepResult.getStepResultSuccess();
+    }
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/service/resource/flight/CreateProjectStep.java:62]: This is repeated a lot with the same meaning. This should be a static function or a class (along with setting the resource as ready).

File: src/main/java/bio/terra/rbs/service/resource/flight/FinishResourceCreationStep.java [hunk 3]
```diff
@@ -27,12 +29,15 @@ public StepResult doStep(FlightContext flightContext) {
     rbsDao.updateResourceAsReady(
         ResourceId.retrieve(flightContext.getWorkingMap()),
         workingMap.get(FlightMapKeys.CLOUD_RESOURCE_UID, CloudResourceUid.class));
+
+    // Having a boolean value in working map to indicate the resource is already marked as READY.
+    // This can prevent a READY resource got rollback.
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/service/resource/flight/FinishResourceCreationStep.java:34]: this comment would go well on central function/class.

File: src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java [hunk 4]
```diff
@@ -40,32 +45,44 @@ public void testCreateGoogleProject_errorDuringProjectCreation() throws Exceptio
...
+    // Verify project and db entity won't get deleted if resource id READY, even the flight fails.
+    TestingFlightFactory.setFlightClassToUse(ErrorAfterCreateResourceFlight.class);
+    FlightManager manager = new FlightManager(flightFactory, stairwayComponent);
+
+    Pool pool = preparePool();
+    rbsDao.createPools(ImmutableList.of(pool));
+    assertTrue(rbsDao.retrieveResources(ResourceState.CREATING, 1).isEmpty());
+    assertTrue(rbsDao.retrieveResources(ResourceState.READY, 1).isEmpty());
+    String flightId = manager.submitCreationFlight(pool).get();
+    blockUntilFlightComplete(flightId);
```
Comment (wchamber) [inline src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java:74]: maybe also assert the flight ended as FAILED? ditto previous test

**Review by wchamber — CHANGES_REQUESTED** [review]

File: src/main/java/bio/terra/rbs/service/resource/flight/CreateProjectStep.java (hunk 2 above)
Comment (yonghaoy) [reply src/main/java/bio/terra/rbs/service/resource/flight/CreateProjectStep.java]: Added a simple helper method in StepUtils, PTAL

File: src/main/java/bio/terra/rbs/service/resource/flight/StepUtils.java [hunk 5]
```diff
@@ -24,4 +28,11 @@ public static void pollUntilSuccess(
               operation.getOperationAdapter().getError().getMessage()));
     }
   }
+
+  /** Check resource is already marked as READY. This can prevent a READY resource got rollback. */
+  public static boolean isResourceReady(FlightContext flightContext) {
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/service/resource/flight/StepUtils.java:33]: I think we should have `void markResourceReady(FlightContext flightContext)` next to it.

**Review by wchamber — APPROVED** [review]

**Pull request merged**